greeter_plugin/greeter_plugin.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+import numpy as np
+import six
+from werkzeug import wrappers
+
+from tensorboard.backend import http_util
+from tensorboard.plugins import base_plugin
+
+class GreeterPlugin(base_plugin.TBPlugin):
+  """A plugin that serves greetings recorded during model runs."""
+
+  # This static property will also be included within routes (URL paths)
+  # offered by this plugin. This property must uniquely identify this plugin
+  # from all other plugins.
+  plugin_name = 'greeter'
+
+  def __init__(self, context):
+    """Instantiates a GreeterPlugin.
+
+    Args:
+      context: A base_plugin.TBContext instance. A magic container that
+        TensorBoard uses to make objects available to the plugin.
+    """
+    # We retrieve the multiplexer from the context and store a reference
+    # to it.
+    self._multiplexer = context.multiplexer
+
+  @wrappers.Request.application
+  def tags_route(self, request):
+    """A route (HTTP handler) that returns a response with tags.
+
+    Returns:
+      A response that contains a JSON object. The keys of the object
+      are all the runs. Each run is mapped to a (potentially empty)
+      list of all tags that are relevant to this plugin.
+    """
+    # This is a dictionary mapping from run to (tag to string content).
+    # To be clear, the values of the dictionary are dictionaries.
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        GreeterPlugin.plugin_name)
+
+    # tagToContent is itself a dictionary mapping tag name to string
+    # content. We retrieve the keys of that dictionary to obtain a
+    # list of tags associated with each run.
+    response = {
+        run: tagToContent.keys()
```

Comment:
nit: `keys()` in python actually returns a `dict_keys` object (and not a list). A `dict_keys` object is not JSON-serializable:

```
Python 3.4.3 (default, Nov 17 2016, 01:08:31) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> d = {1: 2, 3: 4}
>>> d.keys()
dict_keys([1, 3])
>>> import json
>>> json.dumps(d.keys())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: dict_keys([1, 3]) is not JSON serializable
>>> json.dumps(d.keys())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: dict_keys([1, 3]) is not JSON serializable
```

For the sake of python3, I think we should cast to list:  `list(tagToContent.keys())`